crates/gpui/src/elements/text.rs

```diff
@@ -305,11 +305,55 @@ impl TextLayout {
                 }
 
                 let mut line_wrapper = cx.text_system().line_wrapper(text_style.font(), font_size);
-                let text = if let Some(truncate_width) = truncate_width {
-                    line_wrapper.truncate_line(text.clone(), truncate_width, ellipsis)
+                let (text, truncated) = if let Some(truncate_width) = truncate_width {
+                    (
+                        line_wrapper.truncate_line(text.clone(), truncate_width, ellipsis),
+                        true,
+                    )
                 } else {
-                    text.clone()
+                    (text.clone(), false)
                 };
+                if truncated {
+                    //----------------------------------------------
+                    // Case 0: Normal
+                    // Text: abcdefghijkl
+                    // Runs: Run0 { len: 12, ... }
+                    //
+                    // Truncate res: abcd… (truncate_at = 4)
+                    // Run res: Run0 { string: abcd…, len: 7, ... }
+                    // ---------------------------------------------
+                    // Case 1: Drop some runs
+                    // Text: abcdefghijkl
+                    // Runs: Run0 { len: 4, ... }, Run1 { len: 4, ... }, Run2 { len: 4, ... }
+                    //
+                    // Truncate res: abcdef… (truncate_at = 6)
+                    // Runs res: Run0 { string: abcd, len: 4, ... }, Run1 { string: ef…, len:
+                    // 5, ... }
+                    // ----------------------------------------------
+                    // Case 2: Truncate at start of some run
+                    // Text: abcdefghijkl
+                    // Runs: Run0 { len: 4, ... }, Run1 { len: 4, ... }, Run2 { len: 4, ... }
+                    //
+                    // Truncate res: abcdefgh… (truncate_at = 8)
+                    // Runs res: Run0 { string: abcd, len: 4, ... }, Run1 { string: efgh, len:
+                    // 5, ... }, Run2 { string: …, len: 3, ... }
+                    let mut truncate_at = text.len() - ELLIPSIS.len();
```

Comment:
`let truncate_at = text.len().saturating_sub(ELLIPSIS.len());`

Reply:
I think `text.len()` should always be greater than `ELLIPSIS.len()`, since the `text` here is something like `abcdefgh…` which always contains `ELLIPSIS` in it.